---
title: "ITS QC report"
author: "Doctor Ph.D."
date: \today
geometry: margin=3cm
output: 
    pdf_document:
        keep_tex: false
        toc: true
        toc_depth: 3

---

<!-- ========================================================== -->
<!--   Beginning of Preamble : Preamble seldom requires change  -->
<!-- ========================================================== -->

```{r setup, echo=FALSE}
### ================
###   knitr setup
### ================
library(knitr)
opts_chunk$set(
  tidy=FALSE,
  cache=FALSE,
  cache.lazy = FALSE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  dpi=100,
  fig.width=6,
  fig.height=4,
  fig.align = "center",
  dev.args = list(pdf = list(useDingbats = FALSE))
  )

pander::panderOptions("knitr.auto.asis", FALSE)
#this also means that you have to put results='asis' in any block that has pander output
#this lets you loop through variables and produce multiple pander tables and ggplots in a
#single code block!

```

```{r child = 'preamble_quality_trimmed.Rmd'}
```


```{r, Samples error check 1, results='asis'}
### ===========================
###   check for missing samples
### ===========================

### possible issue 1: Samples found in the sample sheet but not in the feature table (0 reads)
s_missing <- s %>%
  filter(!SampleID %in% colnames(counts)) %>%
  select(SampleID, SampleType, isControl, demux_Read_Counts, Read_Counts)

if (any(!s_missing$isControl)) {
  pander(filter(s_missing, !isControl), caption="These samples were in the sample sheet but not in the feature table.", split.table=Inf)
  #stop (simpleError("Please fix"))
}

```


```{r, Samples error check 2}
### possible issue 2: Samples found in the feature table but not in the sample sheet. There must be an error!
in_counts_not_in_s <- setdiff(colnames(counts), s$SampleID)
if (length(in_counts_not_in_s) > 0) {
  stop (simpleError("These SampleID(s) are in the feature table, but not found in the sample sheet.", paste(in_counts_not_in_s, collapse=" ")))
}
```


# Introduction
This report is based on the results of sequencing performed on `r run_date` for `r investigator` Project. 

## DNA content (qPCR) v. read count
```{r fig.width=4, fig.height=8}
s %>%
  ggplot(aes(x=Read_Counts, y=Copy_Numb)) +
  geom_point(shape=21) +
  scale_x_continuous(expand = c(0, 0), trans = "log10") +
  scale_y_continuous(expand = c(0, 0), trans = "log10") +
  theme_clean() +
  ggtitle("Copy Number per ul DNA")

s %>%
  ggplot(aes(x=Read_Counts, y=Ct_ave)) +
  geom_point(shape=21) +
  geom_vline(aes(xintercept = min_reads), lty=2) +
  geom_hline(aes(yintercept = 30), lty=2) +
  scale_x_continuous(expand = c(0, 0), trans = "log10") +

  theme_clean() +
  ggtitle("Average Cycle Number")

```




## Histogram of high quality paired reads per sample
The black dashed vertical line shows the minimum number of reads (`r min_reads`) for analysis. Control samples, if any, were included in the histogram.

```{r reads histogram, fig.width=6, fig.height=4, warning=F}
ggplot(s, aes(x=Read_Counts)) +
    geom_histogram(aes(fill=SampleType), binwidth=0.1, boundary = TRUE) +
    geom_vline(xintercept = min_reads, color="black", linetype="dashed") +
    scale_x_continuous(expand = c(0, 0), trans = "log10") +
    scale_y_continuous(expand = c(0, 0)) +
    theme_classic() +
    theme_bw() + 
    xlab("Number of reads in sample") +
    ylab("Number of samples")
```

\newpage

## Read counts and final library concentration of each sample

```{r, fig.width=10, fig.height=5}

s %>%
  mutate(Read_Counts = ifelse(is.na(Read_Counts), 0, Read_Counts)) %>%
  droplevels() %>%
  ggplot(aes(x = final_library_conc_ng_ul, y = Read_Counts, fill = SampleType)) +
    geom_point(shape = 21) +
    geom_hline(yintercept = min_reads, color="black", linetype="dashed") +
    theme_bw() + 
    scale_y_continuous(trans = "log10") +
    guides(fill = guide_legend(override.aes = list(shape = 21))) +
    labs(y = "Read Counts", x = "Final library DNA conc (ng/uL)")

```


## Average nucleotide quality after demux (before trimming)

```{r nuc_quality_demux}
# remove empty columns (this is a bug that happens sometimes in the fastqc_quality.tsv report)
nsamples=length(unique(s$SampleID))
tmpfqc <- read.delim(pre_fastqc_fp, sep='\t') %>%
  select( where( ~sum(is.na(.x))<nsamples) ) %>%
  pivot_longer(-Samples, names_to="Position", values_to="Quality") %>%
  # mutate(Position = gsub(pattern="[.]", replacement = "-", x=gsub(pattern="X", replacement="", x=Position)))
  group_by(Samples) %>%
    filter(!if_all(Quality, is.na)) %>% # remove empty rows (these should match the samples that had 0 reads at demux)
  ungroup() 
print(paste("2x Sample number:", length(unique(tmpfqc$Samples))))
tmpfqc %>%
  mutate(
    Position = sub("X", "", Position),
    Position = sub("\\.\\d+", "", Position, perl = TRUE),
    Position = as.numeric(Position)) %>%
  mutate(SampleID=sub("^(.*)_R([12])$", "\\1", Samples), Direction=sub("^(.*)_R([12])$", "\\2", Samples)) %>%
  mutate(Direction = paste0("R", Direction)) %>%
  mutate(Direction = factor(Direction)) %>%
  group_by(Direction, Position) %>%
  summarise(MeanQual = mean(Quality), SdQual = sd(Quality)) %>%
  ungroup() %>%
  ggplot(aes(x=Position, y=MeanQual)) + 
    geom_errorbar(aes(ymin=MeanQual - SdQual, ymax=MeanQual + SdQual)) +
    facet_wrap(~ Direction) +
    geom_line() +
    geom_point() +
    theme_clean() +
    labs(x='Position in sequence read', 
         y='Average quality score per sample')
```

## Average nucleotide quality after trimming

```{r nuc_quality_qc}
# remove empty columns (this is a bug that happens sometimes in the fastqc_quality.tsv report)
nsamples=length(unique(s$SampleID))
tmpfqc <- read.delim(post_fastqc_fp, sep='\t') %>%
  select( where( ~sum(is.na(.x))<nsamples) ) %>%
  pivot_longer(-Samples, names_to="Position", values_to="Quality") %>%
  # mutate(Position = gsub(pattern="[.]", replacement = "-", x=gsub(pattern="X", replacement="", x=Position)))
  group_by(Samples) %>%
    filter(!if_all(Quality, is.na)) %>% # remove empty rows (these should match the samples that had 0 reads at demux)
  ungroup() %>%
  mutate(
    Position = sub("X", "", Position),
    Position = sub("\\.\\d+", "", Position, perl = TRUE),
    Position = as.numeric(Position)) %>%
  mutate(SampleID=sub("^(.*)_r.*$", "\\1", Samples), TrimStep=sub("^(.*)_(r.*)_trimmed$", "\\2", Samples), Direction="R1") %>%
  # mutate(Direction = paste0("R", Direction)) %>%
  mutate(TrimStep = factor(TrimStep))

  print(paste("2x Sample number:", length(unique(tmpfqc$Samples))))
tmpfqc %>% 
  group_by(TrimStep, Position) %>%
  summarise(MeanQual = mean(Quality), SdQual = sd(Quality)) %>%
  ungroup() %>%
  ggplot(aes(x=Position, y=MeanQual, color=TrimStep)) + 
    geom_errorbar(aes(ymin=MeanQual - SdQual, ymax=MeanQual + SdQual), position = position_dodge(3)) +
    # facet_wrap(~ TrimStep) +
    scale_color_manual("Pipeline step", values = rev(c("cyan3", "darksalmon"))) +
    geom_line(position = position_dodge(3)) +
    geom_point(position = position_dodge(3)) +
    theme_clean() +
    labs(x='Position in sequence read', 
         y='Average quality score per sample')
```

\newpage
## Library concentration and read counts per plate

The samples are processed on a 96-well plate. The figures illustrate the library concentration, read counts after demultiplexing and read counts after quality control / denoising steps for each well. The gray color represents NA values. 

```{r}
s %>%
  mutate(plate = paste("Plate", plate)) %>%
  ggplot(aes(x=plate_column, y=fct_rev(plate_row), fill = final_library_conc_ng_ul)) +
    geom_tile(fill = "white") +
    geom_point(shape = 21, size = 3) +
    facet_wrap(~plate, ncol = 2) +
    scale_fill_viridis_c() +
    scale_x_continuous(breaks=seq(1,12), expand=c(0,0), limit = c(0, 13), position = "top") +
    theme_bw() +
    theme(
      strip.background = element_blank(),
      strip.placement = "outside",
      panel.grid = element_blank(),
      aspect.ratio = 0.66,
      axis.text = element_text(size = 6)
    ) +
    labs(
      x="",
      y="Plate row",
      fill="Library\nconcentration\n(ng/ul)"
    )
```

```{r}
s %>%
  mutate(plate = paste("Plate", plate)) %>%
  ggplot(aes(x=plate_column, y=fct_rev(plate_row), fill = demux_Read_Counts)) +
    geom_tile(fill = "white") +
    geom_point(shape = 21, size = 3) +
    facet_wrap(~plate, ncol = 2) +
    scale_fill_viridis_c() +
    scale_x_continuous(breaks=seq(1,12), expand=c(0,0), limit = c(0, 13), position = "top") +
    theme_bw() +
    theme(
      strip.background = element_blank(),
      strip.placement = "outside",
      panel.grid = element_blank(),
      aspect.ratio = 0.66,
      axis.text = element_text(size = 6)
    ) +
    labs(
      x="",
      y="Plate row",
      fill="Raw\nread\ncounts"
    )
```


```{r}
s %>%
  mutate(plate = paste("Plate", plate)) %>%
  ggplot(aes(x=plate_column, y=fct_rev(plate_row), fill = Read_Counts)) +
    geom_tile(fill = "white") +
    geom_point(shape = 21, size = 3) +
    facet_wrap(~plate, ncol = 2) +
    scale_fill_viridis_c() +
    scale_x_continuous(breaks=seq(1,12), expand=c(0,0), limit = c(0, 13), position = "top") +
    theme_bw() +
    theme(
      strip.background = element_blank(),
      strip.placement = "outside",
      panel.grid = element_blank(),
      aspect.ratio = 0.66,
      axis.text = element_text(size = 6)
    ) +
    labs(
      x="",
      y="Plate row",
      fill="QC\nread\ncounts"
    )
```

## Quality control trimming steps

Reads kept at primer trimming steps.
```{r reads trim}
qc %>%
  filter(perc_reads>0, initial_reads>0) %>%
  # mutate(SampleType = ifelse(perc_reads==1.00 & trim_stage != "Initial demultiplex", "No reads dropped", as.character(SampleType))) %>%
  ggplot(aes(y = perc_reads, x = trim_stage, group = SampleID)) +
    geom_line(aes(color = SampleType)) +
    geom_point(aes(fill = SampleType), shape = 21) +
    facet_wrap(~direction, ncol = 1) +
    scale_y_continuous(labels = scales::percent) +
    labs(y = "Percent of reads kept", x = "Trimming stage") +
    theme_bw() +
    theme(strip.background = element_blank(),
          panel.grid = element_blank(),
          axis.text.x=element_text(angle=-25, hjust= .1))
```

Number of reads removed after trimming.  
One point per Sample ID. If no reads were removed, no point is shown.
```{r trim zero}

if (any(read_lengths$length ==0)) {
  read_lengths %>%
  filter(length == 0) %>%
  group_by(SampleID, primer) %>%
  mutate(counts = n()) %>%
  ungroup() %>%
  unique() %>%
  ggplot(aes(x = SampleType, y = counts, color = SampleType)) +
    geom_boxplot() +
    ggbeeswarm::geom_quasirandom() +
    theme_bw() + 
    facet_wrap(~primer, ncol = 1) +
    scale_y_continuous(trans = "log10") +
    labs(y = "Number of reads removed after trimming", x = "") +
    theme(strip.background = element_blank(),
          panel.grid = element_blank(),
          axis.text.x=element_text(angle=-25, hjust= .1))
}

```

\newpage

## Most abundant primer sequences matched by vsearch and trimmed
Reverse primer sequence: GCTGCGTTCTTCATCGATGC (GCATCGATGAAGAACGCAGC) \newline
Forward primer sequence: CTTGGTCATTTAGAGGAAGTAA (TTACTTCCTCTAAATGACCAAG)

```{r top three trimmed sequences, results='asis'}

top_trimmed %>%
  pander(split.table=Inf)

```

\newpage

## Whole samples that are above the `r min_reads` read count threshold

```{r table, results='asis'}
pander(table(data.frame(s[, quality_by], factor(ifelse(s$Keep, "Keep", "Discard")))), split.table=Inf)
```



# Taxonomic assignments

```{r heatmap_assignments}

satu_limit <- 0.4
heatmap_file <- paste(Sys.Date(), "_taxonomy_heatmap.pdf", sep='')
heatmap_fp <- here("Output", heatmap_file)

```

Taxonomic assignments were performed using the BROCC program (https://github.com/kylebittinger/brocc)

Heatmap charts were generated from the taxonomic assignments. Each column represents one sample and each row represents one taxon (typically a species). Ranks are included in the plot if the taxon is present in `r 100*prop_cut`% abundance in at least one sample.

The chart is colored white if species were not observed in the sample, dark blue if species were observed at very low abundance.  This allows the reader to quickly survey species presence/absence.  Abundance values exceeding `r 100*satu_limit`% are colored red, indicating an extremely dominant species.

`r if(show.text){paste0("Please see attached plot ", heatmap_file, " for the full heatmap")}`

```{r heatmap time, fig.height=15, fig.width=60}

# select taxa with mean relative abundance of prop_cut in at least one sample type
select_taxa <- summed_props %>%
  as.data.frame() %>% 
  rownames_to_column("Taxa") %>% 
  filter(!grepl("phage|sapiens", Taxa)) %>%
  pivot_longer(-Taxa, names_to="SampleID", values_to="props") %>%
  right_join(s_toPlot, by="SampleID")  %>%
  group_by(SampleType, Taxa) %>%
  mutate(mean_prop = mean(props)) %>%
  ungroup() %>%
  filter(mean_prop > prop_cut) %>%
  mutate(phyla = gsub(" .*", "", Taxa)) %>%
  mutate(phyla = factor(phyla, levels = unique(phyla))) %>%
  mutate(phyla = fct_relevel(phyla, "p__Ascomycota",
                                    "p__Basidiomycota",
                                    "p__Chytridiomycota",
                                    "p__Glomeromycota",
                                    "p__Zygomycota",
                                    "p__Deuteromycota")) %>%
  arrange(phyla) %>%
  pull(Taxa) %>%
  as.character() %>%
  unique()

summed_props[select_taxa, s_toPlot$SampleID] %>%
  pheat() %>%
  pheat_color_saturated(saturation_limit = satu_limit) %>%
  pheat_annotate_cols(s_toPlot[, c("SampleID", quality_by, shape_by)]) %>%
  pheat_display_cols(gaps = factor_gaps(s_toPlot[, quality_by])) %>%
  #using if statement here because ifelse is vectorized
  #have to include brackets around if statement otherwise if will always be true (1 is passed)
  {`if`(show.text, pheat_save(., heatmap_fp), .)}

### top species
# top <- top_table(summed_props[select_taxa, s_toPlot$SampleID], s_toPlot, option=2, prop_cut = prop_cut)
# 
# top$avg_prop <- apply(top,1,mean)
# 
# top <- top %>%
#   rownames_to_column() %>%
#   rename(Taxa = rowname) %>%
#   filter(!row.names(top) %in% c("g__; s__")) %>%
#   mutate(Taxa = gsub("g__","",Taxa)) %>%
#   mutate(Taxa = gsub("; s__"," ", Taxa)) %>%
#   arrange(desc(avg_prop))

```



The chart below is a taxonomic heatmap for the negative and positive control samples. The experimental samples are averaged per sample type. Only the taxa with a mean relative abundance of `r 100*prop_cut`% are shown.

```{r fig.height=8, fig.width=10}

# select taxa with mean relative abundance of 1% in at least one sample type
SampleType_averages <- summed_props %>%
  as.data.frame() %>% 
  rownames_to_column("Taxa") %>% 
  pivot_longer(-Taxa, names_to="SampleID", values_to="props") %>%
  right_join(s_toPlot, by="SampleID")  %>%
  filter(!isControl) %>%
  group_by(SampleType, Taxa, isControl) %>%
  summarize(props = mean(props)) %>%
  ungroup() %>%
  mutate(SampleID=SampleType)

controls <- summed_props %>%
  as.data.frame() %>% 
  rownames_to_column("Taxa") %>% 
  pivot_longer(-Taxa, names_to="SampleID", values_to="props") %>%
  right_join(s_toPlot, by="SampleID")  %>%
  filter(isControl) 

ctrl_df <- bind_rows(SampleType_averages, controls) %>%
  group_by(Taxa) %>%
  mutate(mean_prop = mean(props)) %>%
  ungroup() %>%
  filter(mean_prop > prop_cut) %>%
  mutate(phyla = gsub(" .*", "", Taxa)) %>%
  mutate(phyla = factor(phyla, levels = unique(phyla))) %>%
  mutate(phyla = fct_relevel(phyla, "p__Ascomycota",
                                    "p__Basidiomycota",
                                    "p__Chytridiomycota",
                                    "p__Glomeromycota",
                                    "p__Zygomycota",
                                    "p__Deuteromycota")) %>%
  arrange(phyla) %>%
  select(-phyla) %>%
  mutate(is_control = ifelse(isControl, "Control samples", "Non-control samples")) %>%
  arrange(is_control, SampleID)

ctrl_df %>%
  select(SampleID, Taxa, props) %>%
  pivot_wider(names_from=SampleID, values_from=props) %>%
  pheat(rownames_in="Taxa") %>%
  pheat_annotate_cols(unique(ctrl_df[, c("SampleID", "is_control")])) %>%
  pheat_color_saturated(saturation_limit = satu_limit) %>%
  pheat_display_cols(gaps = factor_gaps(unlist(unique(ctrl_df[ c("SampleID", "is_control")]) %>% select(is_control))))

```

`r if(!have_pos_ctrls){paste0("\\textcolor{red}{\\huge THERE IS NO FUNGI IN THE POSITIVE CONTROLS}")}`

# Relative abundances

## Family level bar plot

Bar plots were generated from the taxonomic assignments. Each column represents one sample and each color represents one family. Only the top 10 families are included in the plot. The rest are lumped into the "Other" category. 

```{r, fig.height = 20, fig.width = 10}
summed_props_f %>% 
  as.data.frame() %>% 
  rownames_to_column("Taxa") %>% 
  pivot_longer(-Taxa, names_to="SampleID", values_to="props") %>%
  right_join(s_toPlot, by="SampleID") %>%
  mutate(Taxa2 = fct_lump(Taxa, 10, w=props)) %>% 
  mutate(Taxa2 = reorder(Taxa2, -props)) %>%
  select(-Taxa) %>%
  group_by(SampleID, Taxa2) %>%
  mutate(props = sum(props)) %>%
  ungroup() %>%
  unique() %>%
  
  ggplot(aes(x=props, y=SampleID, fill=Taxa2)) +
    geom_bar(stat="identity", position = position_fill(reverse = TRUE)) +
    # facet_grid(SampleType + eval(parse(text=color_by))~., space="free", scales="free",
    #            labeller = labeller(.default = start_w_newline, .multi_line = FALSE)) +
      facet_grid(SampleType + isControl~., space="free", scales="free") +
    ggsci::scale_fill_d3(palette = "category20") +
    scale_x_continuous(limits = c(0,1), expand=c(0,0), labels=scales:::percent) +
    theme_classic() +
    theme(
      strip.background = element_blank(),
      strip.text.y = element_text(angle = 0, size = 5)
      ) +
    labs(y="", x="Relative abundance", fill="")
```

\newpage
## Genus level bar plot

Bar plots were generated from the taxonomic assignments. Each column represents one sample and each color represents one genus Only the top 10 genera are included in the plot. The rest are lumped into the "Other" category. 


```{r, fig.height = 20, fig.width = 10}
summed_props_g %>% 
  as.data.frame() %>% 
  rownames_to_column("Taxa") %>% 
  pivot_longer(-Taxa, names_to="SampleID", values_to="props") %>%
  right_join(s_toPlot, by="SampleID") %>%
  mutate(Taxa2 = fct_lump(Taxa, 15, w=props)) %>% 
  mutate(Taxa2 = reorder(Taxa2, -props)) %>%
  select(-Taxa) %>%
  group_by(SampleID, Taxa2) %>%
  mutate(props = sum(props)) %>%
  ungroup() %>%
  unique() %>%
  
  ggplot(aes(x=props, y=SampleID, fill=Taxa2)) +
    geom_bar(stat="identity", position = position_fill(reverse = TRUE)) +
    facet_grid(SampleType + eval(parse(text=color_by))~., space="free", scales="free") +
    ggsci::scale_fill_d3(palette = "category20") +
    scale_x_continuous(limits = c(0,1), expand=c(0,0), labels=scales:::percent) +
    theme_classic() +
    theme(
      strip.background = element_blank(),
      strip.text.y = element_text(angle = 0, size = 5)
      ) +
    labs(y="", x="Relative abundance", fill="")
```


# Beta diversity

## Bray-Curtis distance

### PCoA plot based on Bray-Curtis distance

Here, we use Bray-Curtis distance to compare the species composition of the samples to each other.

The first plot shows the distance between each pair of samples in a single 2D plot.  It is not possible to plot the distances exactly on paper, so we have used a method of ordination called Principal Coordinates Analysis to select the best coordinate system for display.  The percentage of total variance captured along each axis is displayed on the chart.

```{r bray-curtis, fig.height=5, fig.width=7}

s_toPlot %<>%
  filter(Keep) %>%
  filter(!isControl) %>%
  droplevels()

dist_in <- usedist::dist_subset(bc, s_toPlot$SampleID)
plot(make_pcoa_plot(dm = dist_in, s = s_toPlot, color_by=quality_by, shape_by=color_by))

```

\newpage

## Jaccard distance

Here, we use Jaccard distance to compare samples based on shared species membership.  Plots are described above.

### PCoA plot based on Jaccard distance

```{r jaccard, fig.height=5, fig.width=7}

dist_in <- usedist::dist_subset(jd, s_toPlot$SampleID)
plot(make_pcoa_plot(dist_in, s_toPlot,  color_by=quality_by, shape_by=color_by))
```


\newpage

# APPENDIX: Number of reads per plate

```{r, results='asis'}
s %>%
  group_by(plate) %>%
  summarize(Demux = sum(demux_Read_Counts, na.rm=T), QC = sum(Read_Counts, na.rm = T)) %>%
  ungroup() %>%
  pander(split.table=Inf)

s %>%
  group_by(forward_barcode_seq) %>%
  summarize(Demux = sum(demux_Read_Counts, na.rm=T), QC = sum(Read_Counts, na.rm = T), N=n_distinct(SampleID)) %>%
  ungroup() %>%
    mutate(rank=rank(QC)) %>%
  arrange(-rank) %>%
  pander(split.table=Inf)

s %>%
  group_by(reverse_barcode_seq) %>%
  summarize(Demux = sum(demux_Read_Counts, na.rm=T), QC = sum(Read_Counts, na.rm = T), N=n_distinct(SampleID)) %>%
  ungroup() %>%
    mutate(rank=rank(QC)) %>%
  arrange(-rank) %>%
  pander(split.table=Inf)
```
\newpage

Counts of high quality paired reads for each sample
```{r appendix, results='asis'}
s %>% 
  select(SampleID, SampleType, Read_Counts, final_library_conc_ng_ul, Keep) %>%
  rename(library_ng_ul = "final_library_conc_ng_ul") %>%
  arrange(-Read_Counts) %>% 
  pander(split.table=Inf)
```


```{r Generate time stamped report, eval=FALSE, include=FALSE}
#notes
#neat: you can run this following command in the console to give your reports custom names (or date-stamp them)
#rmarkdown::render('Report_shotgun.Rmd',output_file = paste('CEASE.report.', Sys.Date(), '.pdf', sep=''))
library(here)
rstudioapi::documentSaveAll()
dir.create(here::here("output"))

qc_report_fp <- here::here("Scripts", "Rmds", "ITS_QC_Report.Rmd")
qc_output_fp <- here::here("Output", paste('ITS.QC.report.', Sys.Date(), '.pdf', sep=''))

rmarkdown::render(qc_report_fp, output_file = qc_output_fp, "all")

```



